setuptools/dist.py

```diff
@@ -356,19 +356,27 @@ def __init__(self, attrs=None):
         self.src_root = attrs.pop("src_root", None)
         self.patch_missing_pkg_info(attrs)
         self.project_urls = attrs.get('project_urls', {})
+        self.long_description_content_type = attrs.pop(
```

Comment:
I don't think it's a good idea to store this on the `Distribution` instance like this. The way this is structured now, `dist.long_description_content_type` and `dist.metadata.long_description_content_type` could be *two different values*, which is a very confusing API - it's not clear which one is used and why they both persist.

The best options I see here are:

1. The `attrs` version overrides any pre-existing `self.metadata` version.
2. The `self.metadata` overrides any `attrs` version
3. The same as 1 or 2, except that a warning is raised if both are specified.
4. A `long_description_content_type_default` can be specified in `self.metadata`, and that is used if and only if `long_description_content_type` is not specified.

I think 4 is the least ambiguous but it does expand the API so I suppose it shouldn't be done lightly.

@jaraco @benoit-pierre Any thoughts on this?

Reply:
> A long_description_content_type_default can be specified in self.metadata, and that is used if and only if long_description_content_type is not specified.

> I think 4 is the least ambiguous but it does expand the API so I suppose it shouldn't be done lightly.

I'd be happy with this.  I don't know if https://git.openstack.org/cgit/openstack-dev/pbr/tree/pbr/core.py#n106 setting a whole bunch of variables mapped out in https://git.openstack.org/cgit/openstack-dev/pbr/tree/pbr/util.py#n93 is actually an API, as such ... or just something that works well enough.  I suspect the latter; but so far out of all of them, this content-type is the only one that can not be set outside of `setup()` arguments, because it is overwritten unconditionally.

Reply:
The sequence when using PBR is this:

- `setuptools.setup` is called
  - `distutils.core.setup` is called
    - `setuptools.dist.Distribution.__init__` is called
      - `distutils.dist.Distribution.__init__` is called, and the `Distribution.metadata` field is set and populated from `attrs`
        - `Distribution.finalize_options` is called
          - `pbr.core.pbr` is called, overriding part of `Distribution.metadata`
      - missing `Distribution.metadata` fields are filled-in (this is where `long_description_content_type` is overriden even if already set)
    - `setup.cfg` is loaded, overriding parts of `Distribution.metadata`

So `setup.cfg` options override PBR options which in turn override `setup()` options (except for `long_description_content_type`).

I dislike how `project_urls` is currently handled (setting it both at the `Distribution` level and later at the `Distribution.metadata` level), and for this reason I'm not in favor of replicating the behavior with `long_description_content_type`.

If we want to keep:
- using distutils to print a warning for unsupported (by setuptools) options
- but avoid warnings for setuptools specific options or metadata options not supported by distutils
then this is how I would patch the current code:

```diff
 setuptools/dist.py | 35 ++++++++++++++++++++++-------------
 1 file changed, 22 insertions(+), 13 deletions(-)

diff --git i/setuptools/dist.py w/setuptools/dist.py
index 321ab6b7..34267afc 100644
--- i/setuptools/dist.py
+++ w/setuptools/dist.py
@@ -328,6 +328,12 @@ class Distribution(Distribution_parse_config_files, _Distribution):
     distribution for the included and excluded features.
     """
 
+    _DISTUTILS_UNSUPPORTED_METADATA = {
+        'long_description_content_type': None,
+        'project_urls': dict,
+        'provides_extras': set,
+    }
+
     _patched_dist = None
 
     def patch_missing_pkg_info(self, attrs):
@@ -353,25 +359,28 @@ class Distribution(Distribution_parse_config_files, _Distribution):
         self.require_features = []
         self.features = {}
         self.dist_files = []
+        # Filter-out setuptools' specific options.
         self.src_root = attrs.pop("src_root", None)
         self.patch_missing_pkg_info(attrs)
-        self.project_urls = attrs.get('project_urls', {})
         self.dependency_links = attrs.pop('dependency_links', [])
         self.setup_requires = attrs.pop('setup_requires', [])
         for ep in pkg_resources.iter_entry_points('distutils.setup_keywords'):
             vars(self).setdefault(ep.name, None)
-        _Distribution.__init__(self, attrs)
-
-        # The project_urls attribute may not be supported in distutils, so
-        # prime it here from our value if not automatically set
-        self.metadata.project_urls = getattr(
-            self.metadata, 'project_urls', self.project_urls)
-        self.metadata.long_description_content_type = attrs.get(
-            'long_description_content_type'
-        )
-        self.metadata.provides_extras = getattr(
-            self.metadata, 'provides_extras', set()
-        )
+        _Distribution.__init__(self, {
+            k: v for k, v in attrs.items()
+            if k not in self._DISTUTILS_UNSUPPORTED_METADATA
+        })
+
+        # Fill-in missing metadata fields; not supported
+        # by distutils or set by other tools (e.g. pbr).
+        for option, default in self._DISTUTILS_UNSUPPORTED_METADATA.items():
+            for source in self.metadata.__dict__, attrs:
+                if option in source:
+                    value = source[option]
+                    break
+            else:
+                value = default() if default else None
+            setattr(self.metadata, option, value)
 
         if isinstance(self.metadata.version, numbers.Number):
             # Some people apparently take "version number" too literally :)
```

Reply:
@benoit-pierre I am happy with that if you are, and I've translated it into this patch (please re-upload with yourself as author if you wish, I'm not sure how things work around here)

```diff
+            for source in self.metadata.__dict__, attrs:
+                if option in source:
+                    value = source[option]
+                    break
```

One thing I wasn't sure about was above.  I think the intention is for `attrs` to override anything set by `self.metadata`, so we want to walk along the `attrs` entry first.  I swapped it and this is what the test-case enforces; if that's not the intention it will have to be reworked.

I have tested this with the pbr change https://review.openstack.org/565177 and it seems OK

Reply:
> I don't think attrs should override metadata:

OK, rereading your comment I guess I misunderstood that you thought it was wrong.  59ebc0f implements metadata overwriting attrs